### ISSUE TITLE
[spaceship] Skip 2FA pushed token entry (to trusted phone selection), select correct phone no automatically

### ISSUE THREAD

**Issue by janpio** [fastlane/fastlane#13980]
#13487 from @jarrodparkes contained a feature request:

>  I manage Apple developer accounts for ~100 different customers, some organization accounts, some individual accounts. For the individual accounts, I'm forced to access the Apple developer portal using their credentials which means 2FA in most cases. Thankfully, by coordinating with the customer, I can add my phone as a trusted device so that I don't have to contact them each time I need to access their account or generate a session...
> 
> But, the problem arises when I try to generate a web session using fastlane spaceauth -u [apple_id_email]. Since most of these Apple accounts have devices other than my phone registered, fastlane never prompts me to choose my phone for sending the code. Looking at the documentation and this closed issue #10790, there is no way to force fastlane to send the code to a certain number. One option would be to remove all the registered devices from the account, leaving only the desired phone number, but this is problematic because many of the customers I work with are using their devices with Apple throughout the day and logging them out would be BAD.

After #13800 is fixed, @jarrodparkes won't have to use the fallback any more but will be able to exit from the "pushed code entry" to a "select phone number" -> "enter code from text message" flow somehow, but this leaves the need for him to 1. exit the "pushed code entry" and 2. selection of his phone number.

---

We can probably implement an ENV variable to just skip 1. and request the phone numbers.

Then 2. can maybe be skipped with another ENV variable, depends on how much information we get on the phone number if they are very similar (by default Apple return the last 2 digits of a number. But what if you have 2 numbers with identical 3 digits at the end?).

**Comment by janpio** [comment]
#13743 was something completely different, as is this issue which describes a possible useful feature. If you are still having problems or issues with something, please open a new issue so we can have a look at it. Thanks.

**Comment by janpio** [comment]
Started to work on this yesterday.

**Comment by janpio** [comment]
I need your help if you have a trusted phone number(s) in your account:

- Please login at [appleid.apple.com](appleid.apple.com) and select "Didn’t get a verification code?" => "Use Phone Number" instead of entering the pushed 2FA token.
- Then write down the exact string(s) it shows you for your numbers (screenshot is fine as well, copy pasting might be hard because the UI can interpret it as a click)
- Then continue the login, and write down the actual "trusted phone numbers" listed there.
- Finally post the result(s) here (replace the actual digits with `x` if you want - making sure you don't change the amount of digits though!)

The result could look something like this:

```
No 1:
(•••) •••-••66
+1-xxx-xxx-xx66

No 2:
•••• •••••85
+49 xxx xxxxx85 
```

Thanks!

---

Bonus question:
- Do you by chance have 2 phone numbers that have the same last two digits? I want to find out how Apple handles the masking in that case.

**Comment by mariannegru** [comment]
Hi @janpio ,

After login (username/password) I get:
![numbers](https://user-images.githubusercontent.com/1712257/53270818-7176a880-36f5-11e9-8e58-b0830a2926d0.png)
My device is not a trusted device ...

**Comment by mariannegru** [comment]
The API is returning the phone numbers with prefix for example for the one with 95 at the end I get +xx ••••• ••••95.

**Comment by janpio** [comment]
Thanks. Can you also share what the _real_ phone numbers look like after logging in? (Feel free to replace the individual digits (except the last two ones) with `x` to protect your phone numbers).

**Comment by garylai** [comment]
Hi @janpio,

(•••) •••-••09
+1 (xxx) xxx xx09

••• ••• ••44
+64 xxx xxx xx44

••• •••• ••34
+64 xxx xxxx xx34

There are more numbers in my accounts but they are in the same formats as above so I skip those.

**Comment by mariannegru** [comment]
@janpio After login in the Security section I have:
+44 xxxx xxxx95
+353 xx xxx xx43
+40 xxx xxx x54

**Comment by mariannegru** [comment]
If you want to make am implementation to find the phone number what I ddi was to match as much digits as possible from the list, I presume in case of same digits from last 2 and first also from prefix they will provide more digits to be able to select the correct one.

**Comment by janpio** [comment]
Thanks @garylai - that matches what I had above. So it seems they have one formatter for US based numbers (+1) and one for the rest of the world.

@mariannegru Oh, I was not aware of 3 digit country numbers - thanks for including that one!

 >I presume in case of same digits from last 2 and first also from prefix they will provide more digits to be able to select the correct one.

Yeah, that's what I expect as well - but as we have a pretty big community here I thought it might not hurt to ask for an actual example. Maybe someone has that case, and then we have actual confirmation.

If there are no additional answers, I will write the first implementation based on the current knowledge and then see how it goes.

**Comment by janpio** [comment]
There is now a PR at https://github.com/fastlane/fastlane/pull/14436 that introduces a `SPACESHIP_2FA_SMS_DEFAULT_PHONE_NUMBER` environment variable. If it is set, you don't have to manually input `sms` and select a phone number any more, both is done automatically and you should get the token pushed to your trusted phone automatically.
